### DISCUSSION TITLE
Avoid duplicate for CollectionType implementing EntityType in ManyToMany relation

### DISCUSSION THREAD

**Discussion by JuGid** [symfony/symfony#43069] [Q&A] [answered]
Hello,
I've been stuck for about 24h now, sorry if it's a duplicate but I didn't find any answer on Google/Github

Symfony version : **4.4.3**
Problem :  **SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '3-2' for key 'PRIMARY**

I have a simple edit action in my controller (from MakerBundle make:crud). I just changed some names for business reasons : 

```
/**
     * @Route("/edit/{id}", name="profil_edit", methods={"GET","POST"})
     */
    public function profil_edit(Request $request, Profil $profile) : Response
    {
        $form = $this->createForm(ProfilType::class, $profile);
        $form->handleRequest($request);

        if ($form->isSubmitted() && $form->isValid()) {
            $this->getDoctrine()->getManager()->flush();

            return $this->redirectToRoute('profile_index', [], Response::HTTP_SEE_OTHER);
        }

        return $this->render('profils/edit.html.twig',[
            'profil'=>$profile,
            'form'=>$form->createView()
        ]);
    }
``` 

I modified my form to have this builder : 

**Form\ProfilType.php***
```
 public function buildForm(FormBuilderInterface $builder, array $options)
{
    $builder
        ->add('nom')
        ->add('routes', CollectionType::class, [
            'entry_type'=> EntityType::class,
            'allow_add' => true,
            'allow_delete' => true,
            'entry_options'=>[
                'class'=>Route::class,
                'query_builder' => function (EntityRepository $er) {
                    return $er->createQueryBuilder('r')
                        ->orderBy('r.nom', 'ASC');
                },
                'choice_label' => 'nom',
                'choice_value' => 'id'
            ]
        ])
    ;
}

public function configureOptions(OptionsResolver $resolver)
{
    $resolver->setDefaults([
        'data_class' => Profil::class,
    ]);
}
```
My entities : 

**Entity\Profil.php**
```
/**
* @ORM\ManyToMany(targetEntity=Route::class, inversedBy="profiles")
*/
private $routes;

public function __construct()
{
    $this->routes = new ArrayCollection();
}

/**
 * @return Collection|Route[]
 */
public function getRoutes(): Collection
{
    return $this->routes;
}

public function addRoute(Route $route): self
{
    if (!$this->routes->contains($route)) {
        $this->routes[] = $route;
    }

    return $this;
}

public function removeRoute(Route $route): self
{
    $this->routes->removeElement($route);

    return $this;
}
```

**Entity\Route.php**
```
/**
* @ORM\ManyToMany(targetEntity=Profil::class, mappedBy="routes")
*/
private $profiles;

public function __construct()
{
      $this->profiles = new ArrayCollection();
}

/**
 * @return Collection|ProfilInterne[]
 */
public function getProfiles(): Collection
{
    return $this->profiles;
}

public function addProfile(ProfilI $profile): self
{
    if (!$this->profile->contains($rofile)) {
        $this->profiles[] = $profile;
        $profile->addRoute($this);
    }

    return $this;
}

public function removeProfile(Profil $profile): self
{
    if ($this->profiles->removeElement($profile)) {
        $profile->removeRoute($this);
    }

    return $this;
}
```

And my template is just the same template as the [Symfony casts tutorial](https://symfonycasts.com/screencast/collections/add-new-collection-prototype)

The problem is that when I push the submit button to edit my entity through the form with a Route that is already in Profil, I got a duplicate error : 
`SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry '3-2' for key 'PRIMARY'`

I tried to do a DataTransformer, but it's used when doctrine INSERT and not UPDATE (I assume... Because It returns the good ArrayCollection but it's not used). I also tried the preUpdate event, but that didn't work as well.

I don't really know what to use now. Maybe they are calling each other...

Is there a way to prevent from this duplication ? I don't really know how to handle this.

Thanks in advance 😃

**Comment by JuGid** [comment]
Hi,

I still looking for a solution about this problem. If someone has an idea 😄

**Comment by avtarsm** [comment, marked as answer]
Hi,

Any particular reason you are using collection type for many to many relation. ?

Avtar

**Reply by JuGid** [reply]
This is to add multiple entities in one hit. Do you know another good solution ? I'm open to anything to resolve this problem. What I want to do is the same as here : https://symfonycasts.com/screencast/collections/add-new-collection-prototype

**Reply by avtarsm** [reply]
Hi, You will need to do below in case if this is a standard many to many

->add('routes', EntityType::class, [
            'class' => Route::class,
            'multiple' => true,
                'query_builder' => function (EntityRepository $er) {
                    return $er->createQueryBuilder('r')
                        ->orderBy('r.nom', 'ASC');
                },
            
    ])

Avtar

**Reply by JuGid** [reply]
Thanks ! I don't know why this solution didn't come to my mind, I use it everyday ... Thank you so much :)

**Comment by professorbk** [comment]
I have a similar problem:

I have 2 classes with many to many relationship: Foo (owning side) and Bar (inverse side). My form is also CollectionType (of Foo) implementing EntityType (of Bar) (add Bar into Foo). This is some code for my form

FooType:
`->add('Bars', CollectionType::class, array('entry_type' => BarType::class, 'allow_add' => true, 'allow_delete' => true, 'by_reference' => false))`

BarType
```
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('text', EntityType::class,
                array(
                    'class' => Bar::class,
                    'query_builder' => function (EntityRepository $er) {
                        return $er->createQueryBuilder('bar')
                            ->orderBy('bar.text', 'ASC');
                    },
                    'choice_label' => 'text',
                    'by_reference' => false,
                )
            )
        ;
    }
```

- if i use 'multiple' => true, everything works as expected
- if i **dont use** **'multiple' => true** (dont ask me why, this is a requirement from work: add a plus and minus button for the form. When plus button is click, the options from Bar.text are shown, and user can select select only one from those options. When minus button is clicked, remove the selected option), relationships are established but a **duplicate record of selected option is added to Bar** table. 

I would like to ask:
- Why does it happen? Is that the way symfony supposed to works?
- Is there any fix/workaround for it?

I am using Symfony 4.4, doctrine bundle 2.6.3

**Comment by avtarsm** [comment]
Hi,

If you have define the class Bar as Many to Many, then multiple has to be true. In case you need to select only one value from the Bar Type, then it has to be Many To One.

Avtar

**Reply by professorbk** [reply]
@avtarsm thanks for the quick answer!
I would like to clarify this: my goal is to select multible options loaded from Bar->$text. But instead of selecting multiple options in one run (by setting multiple true), i would like to archieve it by adding many times (with a plus button), and each time only one option can be selected.

The reason I am doing this is that i am migrating from many to one to many to many, and I do not want to change the current UI (select one at a time) to multiple select.

If you could tell if there is a workaround to archieve this, I would be appreciated a lot

**Reply by avtarsm** [reply]
@professorbk, in that case you will need to use Many To One. This will be help if you want the user to click on Add button and then selecting a single option from the drop-down.

Hope, i have answered your query.

**Reply by professorbk** [reply]
> @professorbk, in that case you will need to use Many To One. This will be help if you want the user to click on Add button and then selecting a single option from the drop-down.
> 
> Hope, i have answered your query.

But in essense, the relation between my 2 entities is Many to Many, not Many to One. And implementing the form to make it "look like Many to One" causes the duplicate record as i described before.

Do you have any reference that states using Many to One relation is the only/meant to be way to archieve what I want in the form/UI?

**Reply by avtarsm** [reply]
@professorbk,

The reason I stated that you need Many To One, because you had mentioned that you need to select mandatory only one record. In case of Many to Many , you can select one record. If you need to make select only one record as mandatory in Many To Many, it has to be done at the frontend form via javascript. This can be achieved via validation on the form.

**Reply by professorbk** [reply]
> @professorbk,
> 
> The reason I stated that you need Many To One, because you had mentioned that you need to select mandatory only one record. In case of Many to Many , you can select one record. If you need to make select only one record as mandatory in Many To Many, it has to be done at the frontend form via javascript. This can be achieved via validation on the form.

I have been done at the frontend form via javascript already, however, it leads to the duplicated record in the database, which i do not understand if it was wrong at my part or Symfony is not supposed to work that way
